### PR TITLE
Add final-destination per-origin rate quotas

### LINKED ISSUE #742: Add per-origin message and byte-rate quotas at the final logical destination
## Current boundary

The inbound pipeline already has two occupancy bounds:

- node-wide count and retained-byte capacity, partitioned by inbound lane; and
- per-transport-peer count and retained-byte capacity (currently 32 events and 128 MiB).

Those bounds prevent unbounded queue growth, but they are not rate limits. They are also keyed by the immediate transport peer. At a final destination, one signed origin can send through several relay peers and repeatedly refill the logical lanes after earlier work completes. Conversely, charging the last-hop relay as if it were the origin would punish an intermediary for traffic it did not create.

#740 now provides a durable destination-scoped replay window keyed by the verified origin account. The missing layer is a bounded, non-persistent rate quota for new final-destination transactions from that origin.

## Goal

Add message-rate and byte-rate admission per verified origin and inbound lane at the final logical destination.

The quota must prevent one origin, including an origin arriving through several last-hop peers, from monopolizing a lane over time. It complements the existing transport-peer occupancy bounds; it does not replace them.

Conceptual admission order:

1. retain the existing raw transport and per-peer bounds;
2. decode and verify both the outer payload and inner transaction signatures;
3. establish that this node is the final logical destination;
4. atomically compose #740 replay classification with the origin quota decision;
5. reserve logical lane capacity;
6. run application validation and handler effects.

An over-quota transaction must not enter the logical mailbox or reach application validation/dispatch.

## Identity and accounting

Use the #740 destination-scoped identity plus the logical lane:

    OriginQuotaKey = (network_id, origin_account_did, destination_did, inbound_lane)

The origin is recovered from transaction.verification. It is not the delegated session DID, the immediate transport peer, or the outer payload signer. Session rotation therefore does not reset a bucket, and changing relay paths does not create fresh allowance.

Each key owns two pure token buckets:

    OriginQuota {
        messages,
        bytes,
        last_refill,
    }

Both dimensions must admit the transaction. Configuration supplies, per lane:

- message rate and burst;
- byte rate and burst;
- maximum number of active origin/lane records.

Use a monotonic Instant. Sender timestamps and wall-clock adjustments must not affect refill.

The byte cost must be one deterministic logical-message cost derived from the verified input. Define the normal and reassembled-message cases explicitly so chunk envelopes cannot avoid the byte quota or charge one logical transaction multiple times.

## Replay/quota composition

Replay state and quota state have different ownership and must remain separate:

- replay state is durable, fail-closed, and never pressure-evicted;
- quota state is runtime-local, time-dependent, and may evict a fully replenished idle record under pressure.

They may reuse StreamKey/origin extraction, but quota fields must not be serialized into the #740 ReplaySnapshot.

The combined decision must satisfy:

- Replay, Fork, and Stale verdicts do not consume origin tokens.
- Quota rejection does not advance the durable replay window; the same valid sequence may be retried after refill.
- Two concurrent copies of one transaction cannot both pass by racing replay classification and quota admission.
- If replay persistence fails, any provisional quota reservation is rolled back.
- Once both decisions commit, later application validation failure still consumes the quota because the origin caused authenticated application work.

Implement this as one serialized admission boundary or an equivalent reservation/commit protocol. Do not perform two independent check-then-update calls that admit a race.

## State bound and eviction

Quota state must have a hard record bound and no periodic sweeper.

When the table is under pressure:

- a fully replenished idle record may be removed;
- an active or depleted record must not be evicted merely to give the same origin a fresh burst;
- if no safe victim exists, a new key is rejected with a typed capacity verdict.

State removal on connection close is not sufficient because one origin may use several transport peers and reconnecting must not reset its active burst.

## Effects and observability

Quota rejection is a local drop. Do not disconnect the immediate peer: it may be an honest relay carrying another origin's transaction.

Expose typed errors or verdicts for at least:

- message-rate exhaustion;
- byte-rate exhaustion;
- quota-table capacity exhaustion;
- invalid configuration or arithmetic failure.

Publish aggregate counters by lane and reason through the existing measurement surface. Do not use origin DID as an unbounded metrics label.

## Laws

- Under otherwise available node/lane capacity, one origin at or below both configured rates is not rejected by its quota.
- Traffic from origin A cannot consume origin B's tokens.
- The same origin arriving through different relay peers consumes one shared allowance.
- A last-hop relay carrying several origins does not merge their allowances.
- No admitted origin/lane exceeds its configured burst at one instant or its configured long-run message and byte rates.
- Refill and verdicts are deterministic for the same pure state, cost, and monotonic instant.
- Quota state remains within its configured record bound.

## Acceptance criteria

- [ ] Replace this issue's old hash-chain/table assumptions with a standalone pure quota model and documented admission ordering.
- [ ] Add deterministic tests for message and byte refill, burst edges, zero elapsed time, long idle periods, and arithmetic saturation.
- [ ] Show two origins in one lane: one saturating through multiple relay peers does not prevent the other from using its configured allowance.
- [ ] Show one relay carrying multiple origins: one origin's exhaustion does not charge or disconnect the relay or the other origins.
- [ ] Prove Replay, Fork, and Stale inputs do not consume quota.
- [ ] Prove quota rejection does not advance replay state and can be retried after refill.
- [ ] Prove concurrent duplicate delivery cannot pass the combined replay/quota boundary twice.
- [ ] Charge normal and reassembled logical messages exactly once with the documented byte-cost function.
- [ ] Do not create origin quota state on intermediate relay nodes.
- [ ] Bound the origin/lane table, safely reuse fully replenished idle slots, and fail closed when no safe slot exists.
- [ ] Wire conservative per-lane defaults through native and Wasm construction paths.
- [ ] Add aggregate counters for each drop reason without high-cardinality origin labels.
- [ ] Update SECURITY.md to distinguish transport-peer occupancy bounds from final-destination origin rate limits.

## Non-goals

- Replacing the existing per-transport-peer count and memory capacities.
- Rate-limiting signature verification or other pre-origin work; a separate transport-edge rate gate would cover that boundary.
- Disconnecting a relay because an origin behind it exceeds quota.
- Cross-node or globally coordinated quotas.
- Reputation-weighted quotas or DRanking policy (#710-#714).
- Onion queue/session cardinality bounds in crates/node.
- Persisting token balances across restart.

Related: #740 (completed replay boundary), #723 (bounded admission precedent), #689 (measurement surface).

### PR DESCRIPTION
**Pull request by zkjoie**
### Please check if the PR fulfills these requirements

- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

## :large_blue_circle: What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)

Security hardening and inbound admission feature.

## :brown_circle: What is the current behavior? (You can also link to an open issue here)

Inbound transport count/byte bounds limit queued memory per node and immediate peer, but do not rate-limit a verified origin across relay paths. Destination replay protection rejects duplicate sequences but has no runtime-local origin allowance.

## :green_circle: What is the new behavior (if this is a feature change)?

- Adds pure message and byte token buckets keyed by network, verified origin account, final destination, and logical lane.
- Serializes replay classification, quota reservation, and replay persistence as one boundary, including quota rollback on persistence failure.
- Keeps quota state runtime-local and bounded per lane; only fully replenished idle records are safe eviction victims, otherwise admission fails closed.
- Charges normal transactions once from `Transaction.data.len()` and reassembled transactions once after verification, without charging chunk envelopes.
- Enforces quota before logical mailbox admission and application validation while leaving intermediate relays stateless and connected.
- Wires validated defaults through native and Wasm construction paths and exposes bounded aggregate drop counters by lane and reason.

## :radioactive: Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)

No API migration is required. Existing configurations receive conservative per-lane defaults; operators may override the new `origin_quota` section.

## :information_source: Other information

Validated with:

- `cargo +nightly-2026-07-02 fmt --all -- --check`
- `cargo clippy -p rings-core --all-targets --features dummy -- -D warnings`
- `cargo clippy -p rings-node --all-targets -- -D warnings`
- `cargo test -p rings-core --features dummy --quiet` (793 passed, 4 ignored; doc tests 1 passed, 2 ignored)
- `cargo test -p rings-node --quiet` (348 library tests and 6 binary tests passed; 1 ignored)
- `cargo check -p rings-core --target wasm32-unknown-unknown --no-default-features --features wasm`
- `cargo check -p rings-node --target wasm32-unknown-unknown --no-default-features --features browser`

Closes #742

### CODE REVIEW

File: crates/core/src/message/quota.rs [hunk 1]
```diff
@@ -0,0 +1,930 @@
...
+    ) -> Result<Option<(OriginQuotaKey, OriginQuotaInstant)>, OriginQuotaAdmissionError> {
+        let mut victim = None;
+        for (key, quota) in self.records.iter().filter(|(key, _)| key.lane == lane) {
+            if !quota
+                .fully_replenished_at(config, now)
+                .map_err(OriginQuotaAdmissionError::Arithmetic)?
+            {
+                continue;
+            }
+            let candidate = (*key, quota.last_refill);
```
[the remark this answers is not in the document]
Comment (r2d2-rs) [reply crates/core/src/message/quota.rs]: Resolved in 23ffa8bc. Safe victims are ordered by refill instant before key, and the regression now makes the older record have the numerically larger key so the ordering is witnessed.

File: crates/core/src/message/replay.rs [hunk 2]
```diff
@@ -427,59 +461,123 @@ impl TransactionReplay {
...
+            sequence,
+            digest,
+            OriginQuotaLane::Application,
+            0,
+            OriginQuotaInstant::ZERO,
+        )
+        .await
+    }
+
+    #[cfg(test)]
```
[the remark this answers is not in the document]
Comment (r2d2-rs) [reply crates/core/src/message/replay.rs]: Resolved in 23ffa8bc. Both native-only inspection hooks now use the non-Wasm test gate, and the exact CI Wasm Clippy command passes locally.

File: crates/core/src/swarm/transport/tests/test_inbound.rs [hunk 3]
```diff
@@ -17,6 +23,95 @@ mod test_capacity_handoff;
 mod test_pre_admission;
 mod test_storage_interleave;
```
[the remark this answers is not in the document]
Comment (r2d2-rs) [reply crates/core/src/swarm/transport/tests/test_inbound.rs:20]: Resolved in 23ffa8bc. The quota integration scenarios and relay helper moved to test_inbound/test_origin_quota.rs; the parent file is back below 1000 lines.
